### DISCUSSION TITLE
Issue with multiple routes per endpoint with Fastapi >=0.95 (Optional Path Parameter)

### DISCUSSION THREAD

**Discussion by vincentsarago** [fastapi/fastapi#9278] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Query

app = FastAPI()

@app.get("/test")
@app.get("/test/{val}")
def endpoint(val: str = Query("yo")):
    return val
```

results in 
```
# File fastapi/dependencies/utils.py:443, in analyze_param(param_name, annotation, value, is_path_param)
#     441 if field_info is not None:
#     442     if is_path_param:
# --> 443         assert isinstance(field_info, params.Path), (
#     444             f"Cannot use `{field_info.__class__.__name__}` for path param"
#     445             f" {param_name!r}"
#     446         )
#     447     elif (
#     448         isinstance(field_info, params.Param)
#     449         and getattr(field_info, "in_", None) is None
#     450     ):
#     451         field_info.in_ = params.ParamTypes.query

# AssertionError: Cannot use `Query` for path param 'val'
```

### Description

in a lot of my projet I'm wrapping one endpoints with multiple routes and make some of the path parameter optional like in https://github.com/developmentseed/titiler/blob/0245698ba8091feee07d0d25671f853d8be90e5c/src/titiler/core/titiler/core/factory.py#L476-L504

It worked well (🙏) until yesterday release (0.95). I know that what I'm doing might be considered as an edge case but I'm willing to try to fix this if this is considered as a `bug` 🤷 

Thanks for your help 🙏 

first reported over https://github.com/developmentseed/titiler/issues/609

### Operating System

macOS

### Operating System Details

Mac OS / Linux (GitHub actions)

### FastAPI Version

0.95

### Python Version

3.9 (but confirmed in all version)

**Comment by alukach** [comment, marked as answer]
@vincentsarago Looking at your code example, I'm confused why you explicitly use `Query()` for something that can be provided via a  path operator.  If you're vague about where the value comes from, FastAPI will be clever and try to retrieve the value from wherever it can:

```py
from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/test")
@app.get("/test/{val}")
def endpoint(val: str = "yo"):
    return val
```

This way, you can retrieve it from a path or query param (path being preferred):

```
▶ curl "http://127.0.0.1:8000/test"
"yo"%
▶ curl "http://127.0.0.1:8000/test/foo"
"foo"%
▶ curl "http://127.0.0.1:8000/test?val=bar"
"bar"%
▶ curl "http://127.0.0.1:8000/test/foo?val=asdf"
"foo"%
```

**Reply by alukach** [reply]
To fix something like [your linked real-world code usage](https://github.com/developmentseed/titiler/blob/0245698ba8091feee07d0d25671f853d8be90e5c/src/titiler/core/titiler/core/factory.py#L476-L504), I believe you will need to describe the input via a Pydantic model.  This way, you don't need to specify whether something is provided via a path-param or query-param and you can instead leave it to FastAPI to figure out:

```py
    class TileRequest(BaseModel):
        z: int = Field(..., ge=0, le=30, description="TMS tiles's zoom level")
        x: int = Field(..., description="TMS tiles's column")
        y: int = Field(..., description="TMS tiles's row")
        TileMatrixSetId: Literal[tuple(self.supported_tms.list())] = Field(
            self.default_tms,
            description=f"TileMatrixSet Name (default: '{self.default_tms}')",
        )
        scale: int = Field(
            1, gt=0, lt=4, description="Tile size scale. 1=256x256, 2=512x512..."
        )
        format: ImageType = Field(
            None, description="Output image type. Default is auto."
        )
    
    @self.router.get(r"/tiles/{z}/{x}/{y}", **img_endpoint_params)
    @self.router.get(r"/tiles/{z}/{x}/{y}.{format}", **img_endpoint_params)
    @self.router.get(r"/tiles/{z}/{x}/{y}@{scale}x", **img_endpoint_params)
    @self.router.get(r"/tiles/{z}/{x}/{y}@{scale}x.{format}", **img_endpoint_params)
    @self.router.get(r"/tiles/{TileMatrixSetId}/{z}/{x}/{y}", **img_endpoint_params)
    @self.router.get(
        r"/tiles/{TileMatrixSetId}/{z}/{x}/{y}.{format}", **img_endpoint_params
    )
    @self.router.get(
        r"/tiles/{TileMatrixSetId}/{z}/{x}/{y}@{scale}x", **img_endpoint_params
    )
    @self.router.get(
        r"/tiles/{TileMatrixSetId}/{z}/{x}/{y}@{scale}x.{format}",
        **img_endpoint_params,
    )
    def tile(
        tile_req: TileRequest
        # ...
```

**Reply by vincentsarago** [reply]
@alukach (as mentioned in slack) we want to have Optional Path parameter (because we use multiple route) with sometime also defaults. 
Our implementation was following a comment by @tiangolo about Path with default values https://github.com/tiangolo/fastapi/issues/945#issuecomment-585993501

**Reply by vincentsarago** [reply]
>  If you're vague about where the value comes from, FastAPI will be clever and try to retrieve the value from wherever it can:

yeah it's nice but then you loose the nice Path or Query features 🤷 but I guess I don't have many choice (this will force us to do custom openapi docs to set the parameter description 😭 ) 

The `TileRequest` pydantic class in the other end might be nice but we ofter create the dependency dynamically so we would have to create the class dynamically too.

**Reply by alukach** [reply]
> we would have to create the class dynamically too

Yeah, I think dynamically creating a pydantic request model will be a necessary evil to get all the niceties of validation/documentation that comes with `Path()` or `Query()`

**Comment by ebottos94** [comment]
Hi @vincentsarago you can do simply something like this : 

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/test/{val}")
@app.get("/test/")
async def endpoint(val: str|None = "yo"):
    return val
```

**Reply by vincentsarago** [reply]
Thanks @ebottos94, 

as I mentioned in one of the comment

> yeah it's nice but then you loose the nice Path or Query features 🤷 but I guess I don't have many choice (this will force us to do custom openapi docs to set the parameter description 😭 )

to me the documentation `lost` is a bit painful but I guess using `Query` for *optional path parameter* was a *hack*

**Reply by vincentsarago** [reply]
I gave a tried using `Annotated` types (which is similar to using direct type as proposed)
```python
from typing import Annotated
from fastapi import FastAPI

app = FastAPI()

# Note: in Annotated[str, "username"] , `username` is a placeholder 
# and not used anywhere. FastAPI will only look for `Field` (e.g Path, Query) to get the metadata from.
@app.get("/users/{user_id}")
@app.get("/users")
def users(user_id: Annotated[str, "username"] = None): 
    return user_id or "NotVincent"

from starlette.testclient import TestClient
with TestClient(app) as client:
    print(client.get("/users/vincent").json())
    print(client.get("/users").json())

>> Vincent
>> NotVincent
```

I'm now thinking that supporting `string` annotated Metadata to set the OpenAPI description could solve all my issues 🙈

**Comment by larashores** [comment]
I'm seeing a similar issue with the `Annotated` syntax. I need to use `Query` to make sure the OpenAPI generates. This works when either the `/a` or `/b` endpoint is deleted. Is this supposed to be supported?

```python
from typing import Annotated

import fastapi

router = fastapi.APIRouter()

@router.get("/b")
@router.get("/a")
def endpoint(q: Annotated[str | None, fastapi.Query(description="OpenAPI Docs")] = None):
    print(q)
```

```
Traceback (most recent call last):
  File "/Users/lara/coalition/scratchpad/src/python/fastapi_bug.py", line 9, in <module>
    def endpoint(q: Annotated[str | None, fastapi.Query(title="The query parameter")] = None):
  File "/Users/lara/.local/share/virtualenvs/scratchpad-PbginYRt/lib/python3.10/site-packages/fastapi/routing.py", line 661, in decorator
    self.add_api_route(
  File "/Users/lara/.local/share/virtualenvs/scratchpad-PbginYRt/lib/python3.10/site-packages/fastapi/routing.py", line 600, in add_api_route
    route = route_class(
  File "/Users/lara/.local/share/virtualenvs/scratchpad-PbginYRt/lib/python3.10/site-packages/fastapi/routing.py", line 446, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
  File "/Users/lara/.local/share/virtualenvs/scratchpad-PbginYRt/lib/python3.10/site-packages/fastapi/dependencies/utils.py", line 303, in get_dependant
    type_annotation, depends, param_field = analyze_param(
  File "/Users/lara/.local/share/virtualenvs/scratchpad-PbginYRt/lib/python3.10/site-packages/fastapi/dependencies/utils.py", line 387, in analyze_param
    assert field_info.default is Undefined or field_info.default is Required, (
AssertionError: `Query` default value cannot be set in `Annotated` for 'q'. Set the default value with `=` instead.
```

**Comment by MrCreosote** [comment]
The multiple routes per method idea might be a red herring - the problem can be reproduced by sharing a `Query` declaration between two different methods (which I do all the time in my pre-0.95 code):

```python
from typing import Annotated

import fastapi

router = fastapi.APIRouter()

QUERY = fastapi.Query(description="OpenAPI Docs")


@router.get("/b")
def endpoint1(q: Annotated[str | None, QUERY] = None):
    print(q)


@router.get("/a")
def endpoint2(q: Annotated[str | None, QUERY] = None):
    print(q)
```

Results in:
```
In [1]: from src import test

*snip a huge stack trace*

AssertionError: `Query` default value cannot be set in `Annotated` for 'q'. Set the default value with `=` instead.
```
Commenting out `endpoint2` results in a successful import.

**Comment by MrCreosote** [comment]
This might be the fix: https://github.com/tiangolo/fastapi/pull/9315
